#159987418 user can report an article

IyiKuyoro opened the pull request:
#### What does this PR do?
Implement user should be able to report articles
#### Description of Task to be completed?
- Seed new users
- Seed Categories
- Seed articles
- Implement article rating feature
- Implement get all reported cased
#### How should this be manually tested?
1. Clone the repository.
2. Checkout the `ft-report-article-#159987418` branch.
3. Run npm install on your local machine.
4. Run the following commands 
```export NODE_ENV=development
export DATABASE_URL=(Your-Local-DB-URL)
export TEST_DATABASE_URL=(Your-Local-Test-DB-URL)
npm run migrate
sequelize db:seed:all
```
5. Start the server using `npm run start_dev`.
6. Signup using `/api/v1/users/auth/login` and save the token
7. Using the token access the following routes with postman with a post request; `localhost:8000/api/v1/articles/1/report`.
8. After reporting article 1, get all reported cases by accessing `localhost:8000/api/v1/articles/cases`.
#### Any background context you want to provide?
This feature should enable users to report articles and get all report cases.
#### What are the relevant pivotal tracker stories?
#159987418

IyiKuyoro committed:
feat(articles): add test for article reporting> - test for succesful report> - test for article not found> - test for invalid violation> - invalid request body

IyiKuyoro committed:
feat(articles): add `articleRoutes.js` file

> - add file `articleRoutes.js`
> - add basic implementation required

IyiKuyoro committed:
feat(articles): add `articleController.js`

> - add controller functionality on `articleController.js
> - modify test to cover edge case of user rating own article

IyiKuyoro committed:
feat(articles): add `reportValidation.js`> - add report validation middlewares> - modified the order of calling middlewares> - modify test to accomodate changes in functionality

IyiKuyoro committed:
feat(articles): add the cases table to the DB

> - add a cases migration file
> - add a cases model file
> - add node environment script to `start_dev`

IyiKuyoro committed:
feat(article): modify feature and test cased

IyiKuyoro committed:
feat(articles): document the create article endpoint

IyiKuyoro committed:
feat(articles): get all reported cases> - add data seeders> - get all reported cased> - document this feature[Finishes #159987418]

IyiKuyoro committed:
feat(articles): get all reported cases

> - add data seeders
> - get all reported cased
> - document this feature

[Finishes #159987418]

iAmao commented on server/controllers/articleController.js:
```diff
@@ -0,0 +1,38 @@
+      }
+    }).spread((user, created) => {
+      if (!created) {
+        return res.status(400).jsend.fail({
```
Use status code `409` here. it means `Conflict`

iAmao reviewed (commented):
👍 Nice, but let us have a separate resource for `cases`.

We'll have something like:
[GET] `/api/v1/cases`
Something like that.
To filter cases by articles, we'll use a query param.
So something like `/api/v1/cases/?articles=id1,id2`

IyiKuyoro committed:
feat(articles): user can get cases for specified articles

IyiKuyoro committed:
feat(cases): rebase develop

> - fix all conflicts
> - add missing documentations
> - refactor code to pass failing test
> - add articleId validation

IyiKuyoro committed:
feat(cases): add [PUT] resolve case

IyiKuyoro committed:
feat(cases): include the author and article info in the case

IyiKuyoro committed:
feat(cases): include the author and article info in the case

IyiKuyoro committed:
Merge branch 'ft-report-article-#159987418' of https://github.com/andela/metis-ah into ft-report-article-#159987418

IyiKuyoro committed:
Merge branch 'ft-report-article-#159987418' of https://github.com/andela/metis-ah into ft-report-article-#159987418

IyiKuyoro committed:
Merge branch 'ft-report-article-#159987418' of https://github.com/andela/metis-ah into ft-report-article-#159987418

IyiKuyoro committed:
Merge branch 'ft-report-article-#159987418' of https://github.com/andela/metis-ah into ft-report-article-#159987418

IyiKuyoro committed:
resolve all requested changes

IyiKuyoro committed:
resolve all requested changes

IyiKuyoro committed:
Resolve merged conflicts

IyiKuyoro committed:
Resolve merged conflicts

iAmao commented on server/controllers/authController.js:
```diff
@@ -23,7 +23,7 @@ const authController = {
    * @returns {object} The response object
    */
-  returnError: (req, res) => res.status(400).jsend.fail({
+  returnError: (req, res) => res.status(500).jsend.fail({
```
try using `401` error code

iAmao commented on server/controllers/casesController.js:
```diff
@@ -0,0 +1,69 @@
+    let searched = {};
+
+    // Check if a query parameter was passed and parse it into an array
+    if (req.query.a) {
```
use `articles` for query string

iAmao commented on server/helpers/helpers.js:
```diff
@@ -149,6 +151,19 @@ const helpers = {
+   * @returns {array} An array of integers
+   */
+  getIntArray: (str) => {
+    const arrayStr = str.split(' ');
```
use comma separated values here

iAmao commented on test/reportArticle.spec.js:
```diff
@@ -0,0 +1,181 @@
+describe('REPORT AN ARTICLE:', () => {
+  const token = generateToken(7200, { id: 5, isVerified: true });
+
+  it('get all reported cased \'no case\'', (done) => {
```
I think this spec description can be better written as
```
should return a descriptive message if there are no cases found.
````

iAmao commented on test/reportArticle.spec.js:
```diff
@@ -0,0 +1,181 @@
+      });
+  });
+
+  it('successful', (done) => {
```
I think this spec description can be better written as
```
should return a success message after reporting a case
```

iAmao commented on test/reportArticle.spec.js:
```diff
@@ -0,0 +1,181 @@
+      });
+  });
+
+  it('report same article', (done) => {
```
I think this spec description can be better written as
```
should return an appropriate error response if the article has already been reported.
```

iAmao reviewed (changes requested):
Nice just the test descriptions 👍
